File: ai/prompts.go
package ai

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

const defaultLLMPromptsPath = "data/templates/llm_prompts.json"

type llmPromptsJSON struct {
	WorldPhenomenaCognition string `json:"world_phenomena_cognition"`
	PlayerNPC               struct {
		PersonaLine            string `json:"persona_line"`
		RoomContextFmt         string `json:"room_context_fmt"`
		SpaceConsistencyRule   string `json:"space_consistency_rule"`
		BehaviorRules          string `json:"behavior_rules"`
		TraditionalChineseRule string `json:"traditional_chinese_rule"`
		IdentityPrefix         string `json:"identity_prefix"`
		SensitivityPrefix      string `json:"sensitivity_prefix"`
		StyleExamplesHeader    string `json:"style_examples_header"`
		StyleExampleBullet     string `json:"style_example_bullet"`
		UserPlainFmt           string `json:"user_plain_fmt"`
		UserWithMemoryFmt      string `json:"user_with_memory_fmt"`
	} `json:"player_npc"`
	NpcToNpc struct {
		TaskIntro              string   `json:"task_intro"`
		ToneAfterWorld         string   `json:"tone_after_world"`
		RulesBlock             string   `json:"rules_block"`
		PlayerPresentNote      string   `json:"player_present_note"`
		UserMessage            string   `json:"user_message"`
		SpeakerLabel           string   `json:"speaker_label"`
		ListenerLabel          string   `json:"listener_label"`
		IdentityPrefix         string   `json:"identity_prefix"`
		ContextLineFmt         string   `json:"context_line_fmt"`
		RoomAtmospherePrefix   string   `json:"room_atmosphere_prefix"`
		RoomTagsPrefix         string   `json:"room_tags_prefix"`
		RecentEventsHeader     string   `json:"recent_events_header"`
		RecentEventBullet      string   `json:"recent_event_bullet"`
		TopicLineFmt           string   `json:"topic_line_fmt"`
		RelationLineFmt        string   `json:"relation_line_fmt"`
		MemoryLineFmt          string   `json:"memory_line_fmt"`
		RoomDescTruncSuffix    string   `json:"room_desc_trunc_suffix"`
		FallbackListenerReply  string   `json:"fallback_listener_reply"`
		CollapseSaySeparators  []string `json:"collapse_say_separators"`
	} `json:"npc_to_npc"`
	Filters struct {
		MetaBadMarkers []string `json:"meta_bad_markers"`
	} `json:"filters"`
}

var (
	llmPromptsMu     sync.RWMutex
	llmPromptsLoaded bool
	cachedLLMPrompts llmPromptsJSON
	llmPromptsPath   = defaultLLMPromptsPath
)

func SetLLMPromptsPathForTest(path string) {
	llmPromptsMu.Lock()
	defer llmPromptsMu.Unlock()
	llmPromptsPath = path
	llmPromptsLoaded = false
	cachedLLMPrompts = llmPromptsJSON{}
}

func loadLLMPrompts() error {
	llmPromptsMu.Lock()
	defer llmPromptsMu.Unlock()
	if llmPromptsLoaded {
		return nil
	}
	raw, err := os.ReadFile(llmPromptsPath)
	if err != nil {
		alt := filepath.Join("..", llmPromptsPath)
		if raw2, err2 := os.ReadFile(alt); err2 == nil {
			raw = raw2
			err = nil
		}
	}
	if err != nil {
		log.Printf("[ai] llm_prompts.json: %v (path=%q)", err, llmPromptsPath)
		llmPromptsLoaded = true
		return err
	}
	var root llmPromptsJSON
	if err := json.Unmarshal(raw, &root); err != nil {
		log.Printf("[ai] llm_prompts.json parse: %v", err)
		llmPromptsLoaded = true
		return err
	}
	cachedLLMPrompts = root
	llmPromptsLoaded = true
	return nil
}

func prompts() *llmPromptsJSON {
	_ = loadLLMPrompts()
	llmPromptsMu.RLock()
	defer llmPromptsMu.RUnlock()
	return &cachedLLMPrompts
}

func WorldPhenomenaCognitionPrompt() string {
	s := strings.TrimSpace(prompts().WorldPhenomenaCognition)
	if s == "" {
		return ""
	}
	if !strings.HasSuffix(s, "\n") {
		s += "\n"
	}
	return s
}

func playerNPCPersonaLine() string     { return prompts().PlayerNPC.PersonaLine }
func playerNPCRoomContextFmt() string { return prompts().PlayerNPC.RoomContextFmt }
func playerNPCSpaceConsistencyRule() string {
	return prompts().PlayerNPC.SpaceConsistencyRule
}
func playerNPCBehaviorRules() string { return prompts().PlayerNPC.BehaviorRules }
func playerNPCTraditionalRule() string { return prompts().PlayerNPC.TraditionalChineseRule }
func playerNPCIdentityPrefix() string  { return prompts().PlayerNPC.IdentityPrefix }
func playerNPCSensitivityPrefix() string {
	return prompts().PlayerNPC.SensitivityPrefix
}
func playerNPCStyleExamplesHeader() string { return prompts().PlayerNPC.StyleExamplesHeader }
func playerNPCStyleExampleBullet() string  { return prompts().PlayerNPC.StyleExampleBullet }
func playerNPCUserPlainFmt() string        { return prompts().PlayerNPC.UserPlainFmt }
func playerNPCUserWithMemoryFmt() string   { return prompts().PlayerNPC.UserWithMemoryFmt }

func npcToNpcTaskIntro() string         { return prompts().NpcToNpc.TaskIntro }
func npcToNpcToneAfterWorld() string    { return prompts().NpcToNpc.ToneAfterWorld }
func npcToNpcRulesBlock() string        { return prompts().NpcToNpc.RulesBlock }
func npcToNpcPlayerPresentNote() string { return prompts().NpcToNpc.PlayerPresentNote }
func npcToNpcUserMessage() string       { return prompts().NpcToNpc.UserMessage }
func npcToNpcSpeakerLabel() string      { return prompts().NpcToNpc.SpeakerLabel }
func npcToNpcListenerLabel() string     { return prompts().NpcToNpc.ListenerLabel }
func npcToNpcIdentityPrefix() string    { return prompts().NpcToNpc.IdentityPrefix }
func npcToNpcContextLineFmt() string    { return prompts().NpcToNpc.ContextLineFmt }
func npcToNpcRoomAtmospherePrefix() string {
	return prompts().NpcToNpc.RoomAtmospherePrefix
}
func npcToNpcRoomTagsPrefix() string      { return prompts().NpcToNpc.RoomTagsPrefix }
func npcToNpcRecentEventsHeader() string  { return prompts().NpcToNpc.RecentEventsHeader }
func npcToNpcRecentEventBullet() string   { return prompts().NpcToNpc.RecentEventBullet }
func npcToNpcTopicLineFmt() string        { return prompts().NpcToNpc.TopicLineFmt }
func npcToNpcRelationLineFmt() string     { return prompts().NpcToNpc.RelationLineFmt }
func npcToNpcMemoryLineFmt() string       { return prompts().NpcToNpc.MemoryLineFmt }
func npcToNpcRoomDescTruncSuffix() string { return prompts().NpcToNpc.RoomDescTruncSuffix }
func npcToNpcFallbackListenerReply() string {
	return prompts().NpcToNpc.FallbackListenerReply
}
func npcToNpcCollapseSaySeparators() []string {
	return prompts().NpcToNpc.CollapseSaySeparators
}

func metaBadMarkers() []string {
	return prompts().Filters.MetaBadMarkers
}

File: src/ai/prompts.rs
// LLM 固定文案載入，對齊 Go ai/prompts.go（data/templates/llm_prompts.json）。

use serde::Deserialize;
use std::fs;
use std::path::Path;
use std::sync::{LazyLock, RwLock};

pub const DEFAULT_LLM_PROMPTS_PATH: &str = "data/templates/llm_prompts.json";

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct PlayerNpcPrompts {
    pub persona_line: String,
    pub room_context_fmt: String,
    pub space_consistency_rule: String,
    pub behavior_rules: String,
    pub traditional_chinese_rule: String,
    pub identity_prefix: String,
    pub sensitivity_prefix: String,
    pub style_examples_header: String,
    pub style_example_bullet: String,
    pub user_plain_fmt: String,
    pub user_with_memory_fmt: String,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct NpcToNpcPrompts {
    pub task_intro: String,
    pub tone_after_world: String,
    pub rules_block: String,
    pub player_present_note: String,
    pub user_message: String,
    pub speaker_label: String,
    pub listener_label: String,
    pub identity_prefix: String,
    pub context_line_fmt: String,
    pub room_atmosphere_prefix: String,
    pub room_tags_prefix: String,
    pub recent_events_header: String,
    pub recent_event_bullet: String,
    pub topic_line_fmt: String,
    pub relation_line_fmt: String,
    pub memory_line_fmt: String,
    pub room_desc_trunc_suffix: String,
    pub fallback_listener_reply: String,
    pub collapse_say_separators: Vec<String>,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct FiltersPrompts {
    pub meta_bad_markers: Vec<String>,
}

/// 與 `llm_prompts.json` 對應（`deny_unknown_fields` 不加，略過 `_說明` 等）。
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct LlmPromptsJson {
    pub world_phenomena_cognition: String,
    pub player_npc: PlayerNpcPrompts,
    pub npc_to_npc: NpcToNpcPrompts,
    pub filters: FiltersPrompts,
}

struct PromptsState {
    path: String,
    loaded: bool,
    cache: LlmPromptsJson,
}

impl PromptsState {
    fn new() -> Self {
        Self {
            path: DEFAULT_LLM_PROMPTS_PATH.to_string(),
            loaded: false,
            cache: LlmPromptsJson::default(),
        }
    }
}

static STATE: LazyLock<RwLock<PromptsState>> = LazyLock::new(|| RwLock::new(PromptsState::new()));

/// 測試用：重設路徑並清空快取（對齊 Go `SetLLMPromptsPathForTest`）。
pub fn set_llm_prompts_path_for_test(path: impl Into<String>) {
    let mut g = STATE.write().unwrap();
    g.path = path.into();
    g.loaded = false;
    g.cache = LlmPromptsJson::default();
}

/// 載入 JSON（冪等；讀檔或解析失敗時仍標記 `loaded`，與 Go 一致）。
pub fn load_llm_prompts() -> anyhow::Result<()> {
    let mut g = STATE.write().unwrap();
    if g.loaded {
        return Ok(());
    }
    let primary = Path::new(&g.path);
    let raw = fs::read_to_string(primary)
        .or_else(|_| fs::read_to_string(Path::new("..").join(&g.path)));
    let raw = match raw {
        Ok(s) => s,
        Err(e) => {
            tracing::warn!("[ai] llm_prompts.json: {} (path={:?})", e, g.path);
            g.loaded = true;
            return Err(e.into());
        }
    };
    match serde_json::from_str::<LlmPromptsJson>(&raw) {
        Ok(root) => {
            g.cache = root;
            g.loaded = true;
            Ok(())
        }
        Err(e) => {
            tracing::warn!("[ai] llm_prompts.json parse: {}", e);
            g.loaded = true;
            Err(e.into())
        }
    }
}

fn read_cache<R>(f: impl FnOnce(&LlmPromptsJson) -> R) -> R {
    let _ = load_llm_prompts();
    let g = STATE.read().unwrap();
    f(&g.cache)
}

/// 世界現象級認知文案（尾端保證換行），對齊 Go `WorldPhenomenaCognitionPrompt`。
pub fn world_phenomena_cognition_prompt() -> String {
    read_cache(|c| {
        let mut s = c.world_phenomena_cognition.trim().to_string();
        if s.is_empty() {
            return String::new();
        }
        if !s.ends_with('\n') {
            s.push('\n');
        }
        s
    })
}

pub fn player_npc_persona_line() -> String {
    read_cache(|c| c.player_npc.persona_line.clone())
}
pub fn player_npc_room_context_fmt() -> String {
    read_cache(|c| c.player_npc.room_context_fmt.clone())
}
pub fn player_npc_space_consistency_rule() -> String {
    read_cache(|c| c.player_npc.space_consistency_rule.clone())
}
pub fn player_npc_behavior_rules() -> String {
    read_cache(|c| c.player_npc.behavior_rules.clone())
}
pub fn player_npc_traditional_rule() -> String {
    read_cache(|c| c.player_npc.traditional_chinese_rule.clone())
}
pub fn player_npc_identity_prefix() -> String {
    read_cache(|c| c.player_npc.identity_prefix.clone())
}
pub fn player_npc_sensitivity_prefix() -> String {
    read_cache(|c| c.player_npc.sensitivity_prefix.clone())
}
pub fn player_npc_style_examples_header() -> String {
    read_cache(|c| c.player_npc.style_examples_header.clone())
}
pub fn player_npc_style_example_bullet() -> String {
    read_cache(|c| c.player_npc.style_example_bullet.clone())
}
pub fn player_npc_user_plain_fmt() -> String {
    read_cache(|c| c.player_npc.user_plain_fmt.clone())
}
pub fn player_npc_user_with_memory_fmt() -> String {
    read_cache(|c| c.player_npc.user_with_memory_fmt.clone())
}

pub fn npc_to_npc_task_intro() -> String {
    read_cache(|c| c.npc_to_npc.task_intro.clone())
}
pub fn npc_to_npc_tone_after_world() -> String {
    read_cache(|c| c.npc_to_npc.tone_after_world.clone())
}
pub fn npc_to_npc_rules_block() -> String {
    read_cache(|c| c.npc_to_npc.rules_block.clone())
}
pub fn npc_to_npc_player_present_note() -> String {
    read_cache(|c| c.npc_to_npc.player_present_note.clone())
}
pub fn npc_to_npc_user_message() -> String {
    read_cache(|c| c.npc_to_npc.user_message.clone())
}
pub fn npc_to_npc_speaker_label() -> String {
    read_cache(|c| c.npc_to_npc.speaker_label.clone())
}
pub fn npc_to_npc_listener_label() -> String {
    read_cache(|c| c.npc_to_npc.listener_label.clone())
}
pub fn npc_to_npc_identity_prefix() -> String {
    read_cache(|c| c.npc_to_npc.identity_prefix.clone())
}
pub fn npc_to_npc_context_line_fmt() -> String {
    read_cache(|c| c.npc_to_npc.context_line_fmt.clone())
}
pub fn npc_to_npc_room_atmosphere_prefix() -> String {
    read_cache(|c| c.npc_to_npc.room_atmosphere_prefix.clone())
}
pub fn npc_to_npc_room_tags_prefix() -> String {
    read_cache(|c| c.npc_to_npc.room_tags_prefix.clone())
}
pub fn npc_to_npc_recent_events_header() -> String {
    read_cache(|c| c.npc_to_npc.recent_events_header.clone())
}
pub fn npc_to_npc_recent_event_bullet() -> String {
    read_cache(|c| c.npc_to_npc.recent_event_bullet.clone())
}
pub fn npc_to_npc_topic_line_fmt() -> String {
    read_cache(|c| c.npc_to_npc.topic_line_fmt.clone())
}
pub fn npc_to_npc_relation_line_fmt() -> String {
    read_cache(|c| c.npc_to_npc.relation_line_fmt.clone())
}
pub fn npc_to_npc_memory_line_fmt() -> String {
    read_cache(|c| c.npc_to_npc.memory_line_fmt.clone())
}
pub fn npc_to_npc_room_desc_trunc_suffix() -> String {
    read_cache(|c| c.npc_to_npc.room_desc_trunc_suffix.clone())
}
pub fn npc_to_npc_fallback_listener_reply() -> String {
    read_cache(|c| c.npc_to_npc.fallback_listener_reply.clone())
}
pub fn npc_to_npc_collapse_say_separators() -> Vec<String> {
    read_cache(|c| c.npc_to_npc.collapse_say_separators.clone())
}

pub fn meta_bad_markers() -> Vec<String> {
    read_cache(|c| c.filters.meta_bad_markers.clone())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn loads_repo_llm_prompts() {
        let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        let p = root.join(DEFAULT_LLM_PROMPTS_PATH);
        if !p.is_file() {
            return;
        }
        set_llm_prompts_path_for_test(p.to_string_lossy().to_string());
        load_llm_prompts().expect("parse");
        let w = world_phenomena_cognition_prompt();
        assert!(!w.is_empty());
        assert!(w.ends_with('\n'));
        assert!(!player_npc_persona_line().is_empty());
        assert!(!meta_bad_markers().is_empty());
    }
}
